hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedMetadataSyncMetadataWriter.java

```diff
@@ -0,0 +1,144 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class SparkHoodieBackedMetadataSyncMetadataWriter extends SparkHoodieBackedTableMetadataWriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkHoodieBackedMetadataSyncMetadataWriter.class);
+  private final String sourceBasePath;
+  private String inflightInstantTimestamp;
+
+  public SparkHoodieBackedMetadataSyncMetadataWriter(Configuration hadoopConf, HoodieWriteConfig writeConfig,
+                                              HoodieFailedWritesCleaningPolicy failedWritesCleaningPolicy, HoodieEngineContext engineContext,
+                                              String inflightInstantTimestamp, String sourceBasePath) {
+    super(hadoopConf, writeConfig, failedWritesCleaningPolicy, engineContext, Option.of(inflightInstantTimestamp));
+    this.sourceBasePath = sourceBasePath;
+    this.inflightInstantTimestamp = inflightInstantTimestamp;
+  }
+
+  public static HoodieTableMetadataWriter create(Configuration conf,
+                                                 HoodieWriteConfig writeConfig,
+                                                 HoodieFailedWritesCleaningPolicy failedWritesCleaningPolicy,
+                                                 HoodieEngineContext context,
+                                                 String inflightInstantTimestamp) {
+    return new SparkHoodieBackedMetadataSyncMetadataWriter(
+        conf, writeConfig, failedWritesCleaningPolicy, context, inflightInstantTimestamp, writeConfig.getMetadataConfig().getBasePathOverride());
+  }
+
+  protected boolean initializeIfNeeded(HoodieTableMetaClient dataMetaClient,
+                                       Option<String> inflightInstantTimestamp) throws IOException {
+    // Do not initialize the metadata table during metadata sync
+    metadataMetaClient = initializeMetaClient();
+    this.metadata = new HoodieBackedTableMetadata(engineContext, dataWriteConfig.getMetadataConfig(), dataWriteConfig.getBasePath(), true);
+    return true;
+  }
+
+  public void bootstrap(Option<String> boostrapUntilInstantOpt) throws IOException {
+    if (!boostrapUntilInstantOpt.isPresent()) {
+      return;
+    }
+
+    String lastInstantTimestamp = boostrapUntilInstantOpt.get();
+    boolean filesPartitionAvailable = dataMetaClient.getTableConfig().isMetadataPartitionAvailable(MetadataPartitionType.FILES);
+
+    // Check and then open the metadata table reader so FILES partition can be read during initialization of other partitions
+    // initMetadataReader();
+    // Load the metadata table metaclient if required
+    if (dataMetaClient == null) {
+      dataMetaClient = HoodieTableMetaClient.builder().setConf(engineContext.getHadoopConf().get()).setBasePath(metadataWriteConfig.getBasePath()).build();
+    }
+
+    // initialize metadata writer
+    List<DirectoryInfo> partitionInfoList = listAllPartitionsFromFilesystem(lastInstantTimestamp, sourceBasePath);
```

Comment:
if metadata table is enabled in the source table, shouldn't we be calling listAllPartitionsFromMDT(...)